### PR TITLE
Rock-Paper-Scissor pallet implementation

### PR DESCRIPTION
**Pull request by DidacSF**
Implementation of a game of Rock-Paper-Scissors using the Substrate node [template](https://github.com/substrate-developer-hub/substrate-node-template)

The pallet has the following extrinsic methods:

* **create_challenge**: Allows an account to create a new RPS Challenge setting a bet amount to be payed to the winner. 
* **enter_challenge**: Allows another account to enter a created RPS Challenge
* **play_challenge**: Allows each of the challenge participants to play their hand (Rock, Paper or Scissors) the actual value is hashed with a secret so that the other user cannot guess what their opponent has played.
* **reveal_challenge_results**: Reveals the results of the challenge once both accounts have played their hands, this requires that both participants send both their hand and secret code in order to validate the originally stored hashes.

Additionally we have the following storage elements:

* **NextBetId**: Use to generate the next identifier when a bet is placed.
* **ChallengeStore**: A StoreMap used to hold the challenges
* **ChallengePlaysStore**: A StorageDoubleMap used to hold the different plays for all the challenges.

The code doesn't currently have neither tests nor benchmarks.

Most of the relevant code can be found in the `pallets/rps/src/lib.rs` file.

### CODE REVIEW

File: pallets/rps/Cargo.toml [hunk 1]
```diff
@@ -28,6 +30,7 @@ sp-runtime = { default-features = false, version = "5.0.0", git = "https://githu
 default = ["std"]
 std = [
 	"codec/std",
+	"sp-io/std",
```
Comment (cowboy-bebug) [inline pallets/rps/Cargo.toml:34]: `sp-std/std` should be added here as well :)
(most of the Substrate dependencies should be under `std` feature to ensure we can build for `no_std` wasm)

File: pallets/rps/src/lib.rs [hunk 2]
```diff
@@ -14,89 +13,427 @@ mod tests;
 #[cfg(feature = "runtime-benchmarks")]
 mod benchmarking;
 
+pub type ChallengeId = u64;
+
+pub type ChallengePlayHash = [u8; 8];
+
+#[derive(Clone, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub struct OpenChallenge<AccountId: PartialEq + Clone, Balance> {
```
Comment (cowboy-bebug) [inline pallets/rps/src/lib.rs:21]: I like that `OpenChallenge`, `AcceptedChallenge` and `FinishedChallenge` are explicitly separated but they seem to share some fields together.

Can we combine them into one struct - something like:
```rust
pub struct Challenge<AccountId: PartialEq + Clone, Balance> {
	challenger: AccountId,
	rival: Option<AccountId>, // for opening a challenge
	bet_amount: Balance,
	winner: Option<AccountId>,
}
```

File: pallets/rps/src/lib.rs [hunk 3]
```diff
@@ -14,89 +13,427 @@ mod tests;
...
+			let challenge_state =
+				ChallengeState::Open(OpenChallenge { challenger: challenger.clone(), bet_amount });
+
+			NextBetId::<T>::mutate(|x| *x += 1);
+
+			ChallengeStore::<T>::insert(&challenge_id, challenge_state);
+
+			Self::deposit_event(Event::ChallengeCreated(challenge_id, challenger, bet_amount));
+
+			Ok(().into())
```
Comment (cowboy-bebug) [inline pallets/rps/src/lib.rs]: This one can just be:
```suggestion
			Ok(())
```

[DispatchResult is just Result<(), sp_runtime::DispatchError>](https://github.com/paritytech/substrate/blob/25417f5d3c2cf191771f33b738a96eebdcde23a7/frame/support/src/dispatch.rs#L52)

File: pallets/rps/src/lib.rs [hunk 4]
```diff
@@ -14,89 +13,427 @@ mod tests;
...
+		}
+
+		#[pallet::weight(10_000)]
+		pub fn enter_challenge(origin: OriginFor<T>, challenge_id: ChallengeId) -> DispatchResult {
+			let rival = ensure_signed(origin)?;
+
+			if let Err(e) = ChallengeStore::<T>::try_mutate(&challenge_id, |challenge_entry| {
```
Comment (cowboy-bebug) [inline pallets/rps/src/lib.rs]: This `if` block can be removed as the `Error::<T>::*` is mapped to `sp_runtime::DispatchError` (so no need to convert it)
```rust
ChallengeStore::<T>::try_mutate(&challenge_id, |challenge_entry| {
    // your logic that returns either Ok(()) or Error::<T>::SomePalletError
})?;
```

File: pallets/rps/src/lib.rs [hunk 5]
```diff
@@ -14,89 +13,427 @@ mod tests;
...
+		#[pallet::weight(10_000)]
+		pub fn play_challenge(
+			origin: OriginFor<T>,
+			challenge_id: ChallengeId,
+			challenge_play: ChallengePlay,
+			challenger_secret: u64,
+		) -> DispatchResult {
+			let player = ensure_signed(origin)?;
+
+			if let Some(challenge) = ChallengeStore::<T>::get(&challenge_id) {
```
Comment (cowboy-bebug) [inline pallets/rps/src/lib.rs]: Here it's more idiomatic to write:
```suggestion
			let challenge = ChallengeStore::<T>::get(&challenge_id).ok_or(Error::<T>::ChallengeNotFound)?;
```

File: pallets/rps/src/lib.rs [hunk 6]
```diff
@@ -14,89 +13,427 @@ mod tests;
...
+			origin: OriginFor<T>,
+			origin_hand: ChallengePlay,
+			origin_secret: u64,
+			rival_hand: ChallengePlay,
+			rival_secret: u64,
+			challenge_id: ChallengeId,
+		) -> DispatchResult {
+			let player = ensure_signed(origin)?;
+
+			if let Err(e) = ChallengeStore::<T>::try_mutate(&challenge_id, |challenge_entry| {
```
Comment (cowboy-bebug) [inline pallets/rps/src/lib.rs]: You can do the same refactor here

File: .idea/vcs.xml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cowboy-bebug) [inline .idea/vcs.xml]: This file could be removed if it's not required

**Review by cowboy-bebug — APPROVED** [review]
LGTM!
The next step is to write some unit tests. There are lots of examples inside crates under [paritytech/substrate/frame](https://github.com/paritytech/substrate/tree/polkadot-v0.9.17/frame) - see `mocks.rs` and `tests.rs` files by convention.

- `mocks.rs` will have `construct_runtime!` mocks to integrate the pallet into mock runtime
- `tests.rs` contains unit tests run inside `execute_with()` closure

File: pallets/rps/src/lib.rs (hunk 2 above)
Comment (DidacSF) [reply pallets/rps/src/lib.rs:21]: Yeah, the reason I separated them is to make state transitions more explicit and reduce the amount of defensive code I would need to create to validate the state of the structs themselves.

But I can try changing it into a single struct If you prefer it like that.

File: pallets/rps/src/lib.rs (hunk 6 above)
Comment (DidacSF) [reply pallets/rps/src/lib.rs]: Deleted!

**Pull request merged**